json.md

```diff
@@ -0,0 +1,363 @@
+redundant code is impossible to strip away. App sizes cannot be optimized when
+using reflection.
+
+<aside class="alert alert-info" markdown="1">
+**What about dartson?**
+
+The [dartson](https://pub.dartlang.org/packages/dartson) library uses runtime
+reflection, which makes it not compatible with Flutter.
+</aside>
+
+Although we cannot use runtime reflection with Flutter, some libraries give us
+similarly easy to use APIs but are based on code generation instead. This
+approach is covered in more detail in the [code generation
+libraries](#code-generation) section.
+
+<a name="manual-serialization"></a>
+## Serializing JSON manually using dart:convert
+
+Basic JSON serialization in Flutter is very simple. Flutter has a built-in
+`dart:convert` library, which includes a straightforward JSON encoder and
+decoder.
+
+Here is an example JSON for a simple user model.
+
+```json
+{
+  "name": "John Smith",
+  "email": "john@example.com"
+}
+```
+
+With `dart:convert`, we can serialize this JSON model in two ways. Let's have a
+look at both.
+
+### Serializing JSON inline
+
+By looking at [the dart:convert JSON
+documentation](https://api.dartlang.org/stable/1.24.3/dart-convert/JsonCodec-class.html),
+we see that we can decode the JSON by calling the `JSON.decode` method, with our
+JSON string as the method argument.
+
+<!-- skip -->
+```dart
+Map<String, dynamic> user = JSON.decode(json);
+
+print('Howdy, ${user['name']}!');
+print('We sent the verification link to ${user['email']}.');
+```
+
+Unfortunately, `JSON.decode()` merely returns a `Map<String, dynamic>`, meaning
+that we do not know the types of the values until runtime. With this approach,
+we lose most of the statically typed language features: type safety,
+autocompletion and most importantly, compile-time exceptions. Our code can
+become instantly more error-prone.
+
+For example, whenever we access the `name` or `email` fields, we could quickly
+introduce a typo. A typo which our compiler does not know of since our entire
+JSON merely lives in a map structure.
+
+### Serializing JSON inside model classes
+
+We can combat the previously mentioned problems by introducing a plain model
+class, which we call `User`. Inside the `User` class, we have:
+
+* a `User.fromJson` constructor, for constructing a new `User` instance from a
+  map structure
+* a `toJson` method, which converts a `User` instance into a map.
+
+This way, the _calling code_ can now have type safety, autocompletion for the
+`name` and `email` fields and compile-time exceptions. If we make typos or treat
+the fields as `int`s instead of `String`s, our app will not even compile,
+instead of crashing on runtime.
+
+**user.dart**
+
+<!-- skip -->
+```dart
+class User {
+  final String name;
+  final String email;
+
+  User(this.name, this.email);
+
+  User.fromJson(Map<String, dynamic> json)
+      : name = json['name'],
+        email = json['email'];
+
+  Map<String, dynamic> toJson() {
```

Comment:
Would probably use `=>` notation here.

```dart
Map<String, dynamic> toJson() => 
  {
   'name'
...
```

Reply:
I can definitely change it, but the original reasoning for me going the long route was that `dartfmt` always formats it like this:

```dart
Map<String, dynamic> toJson() => {
      'name': name,
      'email': email,
    };
```

Which in my eyes didn't look as pretty. :)